pkg/agent/flexvolume/controller.go

```diff
@@ -0,0 +1,341 @@
+}
+
+func newFlexvolumeController(context *clusterd.Context, volumeAttachmentCRDClient rest.Interface, manager VolumeManager) *FlexvolumeController {
+	return &FlexvolumeController{
+		clientset:                  context.Clientset,
+		volumeManager:              manager,
+		volumeAttachmentController: crd.New(volumeAttachmentCRDClient),
+	}
+}
+
+// Attach attaches rook volume to the node
+func (c *FlexvolumeController) Attach(attachOpts AttachOptions, devicePath *string) error {
+
+	namespace := os.Getenv(k8sutil.PodNamespaceEnvVar)
+	node := os.Getenv(k8sutil.NodeNameEnvVar)
+
+	// Name of CRD is the PV name. This is done so that the CRD can be use for fencing
+	crdName := attachOpts.VolumeName
+
+	// Check if this volume has been attached
+	volumeattachObj, err := c.volumeAttachmentController.Get(namespace, crdName)
+	if err != nil {
+		if !errors.IsNotFound(err) {
+			return fmt.Errorf("failed to get volume CRD %s. %+v", crdName, err)
+		}
+		// No volumeattach CRD for this volume found. Create one
+		volumeattachObj = crd.NewVolumeattachment(crdName, namespace, node, attachOpts.PodNamespace, attachOpts.Pod, attachOpts.MountDir, strings.ToLower(attachOpts.RW) == "ro")
+		logger.Infof("Creating Volume attach Resource %s/%s: %+v", volumeattachObj.Namespace, volumeattachObj.Name, attachOpts)
+		err = c.volumeAttachmentController.Create(volumeattachObj)
+		if err != nil {
+			if !errors.IsAlreadyExists(err) {
+				return fmt.Errorf("failed to create volume CRD %s. %+v", crdName, err)
+			}
+			// Some other attacher beat us in this race. Kubernetes will retry again.
+			return fmt.Errorf("failed to attach volume %s. Volume is already attached by a different pod", crdName)
+		}
+	} else {
+		// Volume has already been attached.
+		// Check if there is already an attachment with RW.
+		index := getPodRWAttachmentObject(volumeattachObj)
+		if index != -1 {
+			// check if the RW attachment is orphaned.
+			attachment := &volumeattachObj.Attachments[index]
+			_, err := c.clientset.Core().Pods(attachment.PodNamespace).Get(attachment.PodName, metav1.GetOptions{})
+			if err != nil {
+				if !errors.IsNotFound(err) {
+					return fmt.Errorf("failed to get pod CRD %s/%s. %+v", attachment.PodNamespace, attachment.PodName, err)
+				}
+
+				// Attachment is orphaned. Update attachment record and proceed with attaching
+				attachment.Node = node
+				attachment.MountDir = attachOpts.MountDir
+				attachment.PodNamespace = attachOpts.PodNamespace
+				attachment.PodName = attachOpts.Pod
+				attachment.ReadOnly = attachOpts.RW == "ro"
+				err = c.volumeAttachmentController.Update(volumeattachObj)
+				if err != nil {
+					return fmt.Errorf("failed to update volume CRD %s. %+v", crdName, err)
+				}
+			} else {
+				// Attachment is not orphaned. Original pod still exists. Dont attach.
+				return fmt.Errorf("failed to attach volume %s. Volume is already attached by pod %s/%s", crdName, attachment.PodNamespace, attachment.PodName)
```

Comment:
might be interesting to also include the original pod's status here in the error message